### ISSUE TITLE
Significant slowdown caused by REMOTE_STORE_MERGE_RESULTS

### ISSUE THREAD

**Issue by atnak** [graphite-project/graphite-web#1670]
I recently did an update from source and found rendering graphs had significantly slowed down.

I tracked problem down to commit bc18f9730b102cd5ab3ba1409b6760471f4b43ea at [this branching](https://github.com/graphite-project/graphite-web/commit/bc18f9730b102cd5ab3ba1409b6760471f4b43ea#diff-57c0c4ffa18833a2e2ea55aab4309ffcR83), since *Calculate best minimal node set* is no longer entered by default.

Setting `REMOTE_STORE_MERGE_RESULTS = False` in `local_settings.py` resolves the problem.

* Since, I'm not using `CLUSTERS` maybe `REMOTE_STORE_MERGE_RESULTS` should not be effective here?

* Another thought.  Should `REMOTE_STORE_MERGE_RESULTS` be `False` by default?  (I don't know enough about operation with clusters to make the call.)

cc @deniszh

**Comment by obfuscurity** [comment]
Did you include the fix from https://github.com/graphite-project/graphite-web/pull/1666 in your test?

**Comment by atnak** [comment]
That seems to be yes. 

(My code base comes from `pip install git+https://github.com/graphite-project/graphite-web.git` that I did earlier in the day, but looking at `webapp/graphite/readers.py` and `webapp/graphite/readers.py` shows both changes applied.)

**Comment by deniszh** [comment]
That's sucks. I still don't test fix...

**Comment by deniszh** [comment]
@atnak 
If you not using clustering, of course, you can disable `REMOTE_STORE_MERGE_RESULTS`.
Still wondering why fix worked for @iksaif ..

**Comment by atnak** [comment]
I did a `pip install git+https://github.com/graphite-project/graphite-web.git --upgrade` just in case and retested, and I can definitely say it's still reproducible at the current master on 1f56f5156aa688c705f0cb60f420a6beddb6cdab.

**Comment by obfuscurity** [comment]
Do you have any graphs that demonstrate the regression? I'd like to compare it against @iksaif's results in https://github.com/graphite-project/graphite-web/pull/1666#issuecomment-243656795.

**Comment by atnak** [comment]
This graph is affected:

```
aliasByNode(sortByTotal(maximumAbove(transformNull(sumSeriesWithWildcards(foo.error_rate.*.error_id.{1,3,9}.point, 4), 0, aggre.error_rate.*.zero), 25)), 2)
```

Metrics are:
```
foo.error_rate.OP.error_id.ERROR_IDS.point
foo.error_rate.OP.zero
```
Where `OP` is one of 2248 identifiers, and `ERROR_IDS` is 6.5 identifiers on averages (max 11).

**Comment by obfuscurity** [comment]
wut

**Comment by iksaif** [comment]
With my fix I still had a ~30% penalty, which was better than the massive slowdown that I got without.
Could you dump the infos logs ? This should tell us more about what's happening. (With my fix I was definitely doing only on request per replica instead of one per metric thanks to the bulk_query batching)

Anything with a wildcard will be affected. A good way to debug is checking how many queries are generated by carbon.agents.*.metricsReceived (should be one per replica).

**Comment by bmhatfield** [comment]
@atnak I believe @obfuscurity was asking for a graph *picture* that demonstrates your problem, since we obviously don't have access to the data your query is referencing.

**Comment by atnak** [comment]
I have a feeling a picture would be just as unhelpful but here goes:

![image](https://cloud.githubusercontent.com/assets/918211/18207034/47494544-7164-11e6-8f68-4b0a609d3dcf.png)

**Comment by atnak** [comment]
> Where OP is one of 2248 identifiers, and ERROR_IDS is 6.5 identifiers on averages (max 11).

This works out to 14612 metrics times data points of (10m:30d) over 8 hours per render.  (But it's actually twice that because I'm doing something inefficient with `foo.error_rate.OP.zero`.  (It's inefficient because I don't actually need a wildcard here I just realized).

**Comment by atnak** [comment]
> Could you dump the infos logs ? This should tell us more about what's happening. (With my fix I was definitely doing only on request per replica instead of one per metric thanks to the bulk_query batching)

I'm sorry I don't have a sandbox to get this.  I'm at work and the server I'm working on gets hit by requests by as soon as I restart uwsgi, making it impossible to separate out lines.

**Comment by obfuscurity** [comment]
I was hoping you could show a graph that visually corroborates your regression claims, similar to what @iksaif did in the comment I linked earlier.

**Comment by deniszh** [comment]
@atnak - what's a slowdown (in seconds) for your graph above was introduced?

Could you please check render time when `REMOTE_STORE_MERGE_RESULTS` enabled and `REMOTE_STORE_MERGE_RESULTS` disabled?

Is is still worse than before if `REMOTE_STORE_MERGE_RESULTS` is disabled?

**Comment by atnak** [comment]
> similar to what @iksaif did in the comment I linked earlier

Right, this looks doable looking at uwsgi.log..

**Comment by obfuscurity** [comment]
@atnak I typically use collectd's tail plugin to log retrieval time as first demonstrated by @jssjr [here](https://gist.github.com/jssjr/7159484).

**Comment by obfuscurity** [comment]
@iksaif My main concern here isn't so much the slowdown for remote fetches as that we're slowing down _at all_ for local fetches. As @atnak mentioned, why should anyone hit this if they're not using clusters?

**Comment by obfuscurity** [comment]
@iksaif Sorry, this isn't your problem; you worked on a performance fix, not the original timeseries merge code.

**Comment by deniszh** [comment]
@obfuscurity - yes, I was also surprised that my patch affects local fetches.

**Comment by atnak** [comment]
Okay, I thought about either familiarizing myself with `collectd` or grokking cache.log manually but I wanted to get some sleep so here's one I took with `cat storage/log/webapp/uwsgi.log | cut -d' ' -f16,25` and plotted in *gasp* Excel.

![image](https://cloud.githubusercontent.com/assets/918211/18212079/0066bcfc-717c-11e6-8510-5da8c1da080c.png)

* At 1:30, REMOTE_STORE_MERGE_RESULTS is `False`
* At 1:40, REMOTE_STORE_MERGE_RESULTS is commented out and uwsgi is restarted
* At 1:50, REMOTE_STORE_MERGE_RESULTS is set back to `False` and uwsgi is restarted

I didn't intentionally make any requests to Graphite during this time.  The plotted render times are purely the result of number of Grafana dashboards open and refreshing automatically every 30 seconds or so.

**Comment by atnak** [comment]
For whatever it's worth:
* @iksaif's fix appears to be [about increasing parallelism](https://github.com/graphite-project/graphite-web/pull/1666/commits/a87035629b3d1fb5a95e655e42265fbcf1dd3f56), and I'm not sure how this can help for local fetches.
    * On the other hand, could it possibly decrease performance for local fetches?  (I'm mainly concerned about the comment in the commit message about not benefiting from reusing queries anymore.)
* Is [this condition](https://github.com/graphite-project/graphite-web/commit/bc18f9730b102cd5ab3ba1409b6760471f4b43ea?w=1#diff-57c0c4ffa18833a2e2ea55aab4309ffcR83) missing a `if not local` the way it's done [here](https://github.com/graphite-project/graphite-web/commit/bc18f9730b102cd5ab3ba1409b6760471f4b43ea?w=1#diff-57c0c4ffa18833a2e2ea55aab4309ffcR40) and [here](https://github.com/graphite-project/graphite-web/commit/bc18f9730b102cd5ab3ba1409b6760471f4b43ea?w=1#diff-57c0c4ffa18833a2e2ea55aab4309ffcR52)?

**Comment by iksaif** [comment]
Could you send us your local_settings.py ? Most important settings are CARBONLINK* and REMOTE* and CLUSTER* variables.

**Comment by atnak** [comment]
```diff
--- webapp/graphite/local_settings.py.example   2016-09-02 22:46:10.000000000 +0900
+++ webapp/graphite/local_settings.py   2016-09-03 01:50:57.070934105 +0900
@@ -11,7 +11,7 @@
 # install. This key is used for salting of hashes used in auth tokens,
 # CRSF middleware, cookie storage, etc. This should be set identically among
 # instances if used behind a load balancer.
-#SECRET_KEY = 'UNSAFE_DEFAULT'
+SECRET_KEY = ...

 # In Django 1.5+ set this to the list of hosts your graphite instances is
 # accessible as. See:
@@ -21,7 +21,7 @@
 # Set your local timezone (Django's default is America/Chicago)
 # If your graphs appear to be offset by a couple hours then this probably
 # needs to be explicitly set to your local timezone.
-#TIME_ZONE = 'America/Los_Angeles'
+TIME_ZONE = 'Japan'

 # Set the default short date format. See strftime(3) for supported sequences.
 #DATE_FORMAT = '%m/%d'
@@ -264,6 +264,9 @@
 # that choosing the "most complete" one (pre-0.9.14 behaviour).
 #REMOTE_STORE_MERGE_RESULTS = True

+# 2016-09-02: this destroys performance
+REMOTE_STORE_MERGE_RESULTS = False
+
 ## Remote rendering settings
 # Set to True to enable rendering of Graphs on a remote webapp
 #REMOTE_RENDERING = True
```

`egrep 'CARBONLINK|REMOTE|CLUSTER' webapp/graphite/local_settings.py`:
```
## REMOTE_USER authentication. See: https://docs.djangoproject.com/en/dev/howto/auth-remote-user/
#USE_REMOTE_USER_AUTHENTICATION = True
#CLUSTER_SERVERS = ["10.0.2.2:80", "10.0.2.3:80"]
#INTRACLUSTER_HTTPS = False
#REMOTE_FIND_TIMEOUT = 3.0           # Timeout for metric find requests
#REMOTE_FETCH_TIMEOUT = 3.0          # Timeout to fetch series data
#REMOTE_RETRY_DELAY = 60.0           # Time before retrying a failed remote webapp
#REMOTE_EXCLUDE_LOCAL = False
#REMOTE_STORE_MERGE_RESULTS = True
REMOTE_STORE_MERGE_RESULTS = False
#REMOTE_RENDERING = True
# access to metric data or should have CLUSTER_SERVERS configured
#REMOTE_RENDER_CONNECT_TIMEOUT = 1.0
#CARBONLINK_HOSTS = ["127.0.0.1:7002:a", "127.0.0.1:7102:b", "127.0.0.1:7202:c"]
#CARBONLINK_TIMEOUT = 1.0
#CARBONLINK_RETRY_DELAY = 15 # Seconds to blacklist a failed remote server
#CARBONLINK_HASHING_KEYFUNC = "/opt/graphite/bin/keyfuncs.py:my_keyfunc"
```

**Comment by atnak** [comment]
>    * On the other hand, could it possibly decrease performance for local fetches?  (I'm mainly concerned about the comment in the commit message about not benefiting from reusing queries anymore.)

Regarding this, I tested it and found that it doesn't make the slightest difference.

The test I carried out is similar to [before](https://github.com/graphite-project/graphite-web/issues/1670#issuecomment-244434540) except runs where compared with changes to `readers.py` applied and not applied.  `REMOTE_STORE_MERGE_RESULTS = False` in both cases.

Here's the uneventful graph:  (Changes in a87035629b3d1fb5a95e655e42265fbcf1dd3f56 to `readers.py` reverted and uwsgi restarted at 20:05)

![image](https://cloud.githubusercontent.com/assets/918211/18224563/b614ea0e-7214-11e6-97a3-2a0370fd3c37.png)

**Comment by deniszh** [comment]
Thanks, @atnak !
Indeed, my patch shouldn't affect local queries.
Back to the drawing board!
![image](https://cloud.githubusercontent.com/assets/1227222/18224705/26ca87e4-71df-11e6-9240-20221f8af652.png)

**Comment by deniszh** [comment]
I'm suspecting that fix should be simple as https://github.com/deniszh/graphite-web/commit/7f3c01b0088f3d8c70881aedfb08865ce709989e, testing it now.

**Comment by atnak** [comment]
> I'm suspecting that fix should be simple as deniszh@7f3c01b, testing it now.

I'm getting second thoughts about this.  What happens if you're using clusters but also using local?  (Or isn't this a possible setup?)  Wouldn't *"we need all nodes for merging results"* apply in that case?

**Comment by deniszh** [comment]
@atnak 
Yes, that's exactly what I'm testing now. And it looks like local merging works too. :)
Preparing PR.

**Comment by deniszh** [comment]
@atnak - could you please test my fix in PR #1674 ?  I tested it only on synthesize - unfortunately I had no production installation on master yet.
Much appreciated!

**Comment by deniszh** [comment]
Fixed in PR #1674 with many help of @atnak
